README.md
# Clippings

Python module to manipulate Amazon Kindle clippings files.

It allows you to:

* Parse your existing clippings file into a structured representation;
* Generate clippings file programmatically.

[![Latest version released on PyPI](https://img.shields.io/pypi/v/clippings.svg)](https://pypi.org/pypi/clippings)
[![Build status](https://github.com/samueldg/clippings/workflows/Run%20tests/badge.svg)](https://github.com/samueldg/clippings/actions)

## Installation

```sh
pip install clippings
```

## Usage

### Command-line Usage

```sh
# Parse a clippings file
clippings -o json ./clippings.txt

# or from stdin:
cat clippings.txt | clippings -
```

### Programmatic Usage

```py
from clippings import parse_clippings

my_clippings_file = ...
parse_clippings(my_clippings_file)
```

### Want to parse non-English clippings?

Here's a highlight clipping taken from a Kindle that speaks Spanish::

```txt
El Principe de la Niebla (Carlos Ruiz Zafón)
- La subrayado en la página 4 | posición 60-60 | Añadido el miércoles, 6 de julio de 2022 06:54:57

asintiendo a una pregunta que Max no había llegado a formular.
```

`parse_clippings` won't parse this by default but you can write your own parser for 
the second line and pass it to the parameter `metadata_parser`.

You can take a look at an example [here](./examples/bilingual_spanish_english_kindle/main.py).


CHANGELOG.md
# Change Log

This document tracks changes to [clippings](https://pypi.org/pypi/clippings) between releases.

## [0.9.0](https://github.com/samueldg/clippings/releases/tag/0.9.0) (2022-11-03)

* [feature] Extend API to allow custom metadata line parsers. Includes an example of providing an English + Spanish bilingual parser. (@jonahsol)
* [feature] Allow document title lines that don't include an author. (@jonahsol)
* [dist] Drop support for Python 3.6.
* [dist] Add support for Python 3.11, and include in CI.
* [misc] Upgrade all dev dependencies.
* [misc] Add linter step in Github Actions CI.

## [0.8.0](https://github.com/samueldg/clippings/releases/tag/0.8.0) (2021-11-09)

* [dist] Add support for Python 3.10.

## [0.7.0](https://github.com/samueldg/clippings/releases/tag/0.7.0) (2021-01-13)

* [fix] Compatibility with latest clippings file format (capitalized "Page") (@rjalexa)
* [misc] Ditch HTTP links in docs, in favor of HTTPS
* [misc] Better qualify dev dependencies
* [dist] Add support for Python 3.9.
* [dist] Drop support for Python 3.5.
* [misc] Switch CI over to GitHub actions.
* [test] Convert tests from unittest style to pytest style.

## [0.6.0](https://github.com/samueldg/clippings/releases/tag/0.6.0) (2020-06-27)

* [feature] Compatibility with Paperwhite 3 clippings format (@twkrol)

## [0.5.0](https://github.com/samueldg/clippings/releases/tag/0.5.0) (2019-11-24)

* [dist] Drop support for Python 2.7.
* [dist] Drop support for Python 3.4.
* [dist] Add support for Python 3.8, and include in CI.

## [0.4.0](https://github.com/samueldg/clippings/releases/tag/0.4.0) (2018-01-08)

* [dist] Officially drop support for Python 3.3.
* [dist] Add support for Python 3.7, and include in CI.
* [fix] Fix Travis CI build, which no longer supported Python 3.3.

## [0.3.0](https://github.com/samueldg/clippings/releases/tag/0.3.0) (2017-01-19)

* [feature] Add a 'clippings' console script.

## [0.2.1](https://github.com/samueldg/clippings/releases/tag/0.2.1) (2017-01-08)

* [dist] Add support for Python 3.6, and include in CI.

## [0.2.0](https://github.com/samueldg/clippings/releases/tag/0.2.0) (2016-11-29)

* [feature] Ability to specify the write file using ``--write-to`` or ``-w``.
* [dist] Officially dropped support for Python 2.6.
* [fix] Print a list/JSON array, rather than concatenated dict's/JSON objects (backwards-incompatible).
* [fix] Removed leading zero's in hours, to align with the Kindle format.
* [fix] Fixed broken Python 2 support (module, and tests).
* [test] Added tests for the output functions and ``main()``.
* [test] Added all supported Python version in test automation.
* [misc] Added ``.coveragerc`` file for coverage configuration.
* [misc] Added coveralls to CI process, and badge on the README.

## [0.1.2](https://github.com/samueldg/clippings/releases/tag/0.1.2) (2016-11-23)

* [fix] The clipping page should always return an ``int``.
* [fix] ``authors`` will be a string, not a list (no easy way to parse this reliably).
* [fix] Removed the initial new line from notes' content.
* [test] Added a test suite for utils and parsing classes.
* [test] Integrated coverage and travis CI.
* [misc] Classes implement ``__str__``, and not ``__repr__``.

## [0.1.1](https://github.com/samueldg/clippings/releases/tag/0.1.1) (2016-11-20)

* [dist] Pushed to PyPI!
* [dist] Small fixes for distribution.
* [docs] README converted from markdown into reStructuredText.

## [0.1.0](https://github.com/samueldg/clippings/releases/tag/0.1.0) (2016-11-20)

* [dist] Initial release.


clippings/parser.py
"""Parser for Amazon Kindle clippings file"""
import argparse
import json
import re
from typing import Callable

import dateutil.parser

from clippings.utils import BasicEqualityMixin
from clippings.utils import DatetimeJSONEncoder


DATETIME_FORMAT = '%A, %B %d, %Y %I:%M:%S %p'  # E.g. Friday, May 13, 2016 11:23:26 PM
CLIPPINGS_SEPARATOR = '=========='


class Document(BasicEqualityMixin):
    """Document (e.g. book, article) the clipping originates from.

    A document has a title, and one or multiple authors (in a string).
    """

    PATTERN = re.compile(r'^(?P<title>.+) \((?P<authors>.+?)\)$')

    def __init__(self, title, authors=None):
        self.title = title
        self.authors = authors

    def __str__(self):
        if self.authors:
            return '{title} ({authors})'.format(title=self.title,
                                                authors=self.authors)
        else:
            return self.title

    def to_dict(self):
        return self.__dict__

    @classmethod
    def parse(cls, line):
        match = re.match(cls.PATTERN, line)
        if match:
            return cls(match.group('title'), match.group('authors'))
        else:
            return cls(line, None)


class Location(BasicEqualityMixin):
    """Location of the clipping in the document.

    A location consists of a begin-end range.
    """

    def __init__(self, begin, end):
        self.begin = begin
        self.end = end

    def __str__(self):
        if self.begin == self.end:
            return str(self.begin)
        else:
            return '{0}-{1}'.format(self.begin, self.end)

    def to_dict(self):
        return self.__dict__

    @classmethod
    def parse(cls, string):
        ranges = string.split('-')
        if len(ranges) == 1:
            begin = end = ranges[0]
        else:
            begin = ranges[0]
            end = ranges[1]
        return cls(int(begin), int(end))


class Metadata(BasicEqualityMixin):
    """Metadata about the clipping:

        - The category of clipping (Note, Highlight, or Bookmark);
        - The location within the document;
        - The timestamp of the clipping;
        - The page within the document (not always present).
    """

    PATTERN = re.compile(
        r'^- Your (?P<category>\w+) '
        r'(on|at) ([Pp]age (?P<page>\d+) \| )?'
        r'[Ll]ocation (?P<location>\d+(-\d+)?) \| '
        r'Added on (?P<timestamp>.+)$'
    )

    HOUR_PATTERN = re.compile(r'0(\d:\d{2}:\d{2})')

    def __init__(self, category, location, timestamp, page=None):
        self.category = category
        self.location = location
        self.timestamp = timestamp
        self.page = page

    def __str__(self):
        page_string = '' if self.page is None else 'page {0} | '.format(self.page)

        # Remove leading zero's from the timestamp.
        # They are not present in the Kindle format, but can't be avoided
        # in strftime.
        timestamp_str = self.timestamp.strftime(DATETIME_FORMAT)
        timestamp_str = re.sub(self.HOUR_PATTERN, r'\1', timestamp_str)

        return '- Your {category} on {page}Location {location} | Added on {timestamp}'.format(
            category=self.category.title(),
            page=page_string,
            location=self.location,
            timestamp=timestamp_str,
        )

    def to_dict(self):
        return {
            'category': self.category,
            'location': self.location.to_dict(),
            'timestamp': self.timestamp,
            'page': self.page,
        }

    @classmethod
    def parse(cls, line):
        match = re.match(cls.PATTERN, line)
        category = match.group('category')
        location = Location.parse(match.group('location'))
        timestamp = dateutil.parser.parse(match.group('timestamp'))
        try:
            page = int(match.group('page'))
        except TypeError:
            page = None
        return cls(category, location, timestamp, page)


class Clipping(BasicEqualityMixin):
    """Kindle clipping: content associated with a particular document"""

    def __init__(self, document, metadata, content):
        self.document = document
        self.metadata = metadata
        self.content = content

    def __str__(self):
        return '\n'.join([str(self.document), str(self.metadata), str(self.content)])

    def to_dict(self):
        return {
            'document': self.document.to_dict(),
            'metadata': self.metadata.to_dict(),
            'content': self.content,
        }


def parse_clippings(
        clippings_file,
        document_parser: Callable[[str], Document] = Document.parse,
        metadata_parser: Callable[[str], Metadata] = Metadata.parse):
    """Take a file containing clippings, and return a list of objects."""

    # Last separator not followed by an entry
    entries = clippings_file.read().split(CLIPPINGS_SEPARATOR)[:-1]
    clippings = []

    for entry in entries:
        lines = entry.strip().splitlines()

        document_line = lines[0]
        document = document_parser(document_line)

        metadata_line = lines[1]
        metadata = metadata_parser(metadata_line)

        content = '\n'.join(lines[3:])

        clippings.append(Clipping(document, metadata, content))

    return clippings


def as_kindle(clippings):
    """Return the clippings in the original Kindle format.

    This can useful to programatically create clippings file.
    """
    string = ''
    for clipping in clippings:
        string += '\n'.join([
            str(clipping.document),
            str(clipping.metadata),
            '',
            str(clipping.content),
            CLIPPINGS_SEPARATOR,
            ''
        ])
    return string


def as_dicts(clippings):
    """Return the clippings as python dictionaries.

    The result can be converted to JSON, or manipulated directly in Python,
    for instance.
    """
    return [clipping.to_dict() for clipping in clippings]


def as_json(clippings):
    """Return the clippings as a JSON string."""
    return json.dumps(as_dicts(clippings), cls=DatetimeJSONEncoder)


def main():
    """Read the provided clippings file, parse it,
    then print it using the provided format.
    """
    parser = argparse.ArgumentParser(description='Kindle clippings parser')
    parser.add_argument('file', type=argparse.FileType('r'))
    parser.add_argument('-o', '--output', dest='output',
                        choices=['json', 'dict', 'kindle'], default='json')
    parser.add_argument('-w', '--write-to', dest='write_to', default='-',
                        type=argparse.FileType('w'))
    args = parser.parse_args()

    clippings = parse_clippings(args.file)

    format_functions = {  # Which function to call, depending on 'output' type
        'kindle': as_kindle,
        'dict': as_dicts,
        'json': as_json,
    }
    format_function = format_functions[args.output]
    print(format_function(clippings), file=args.write_to, end='')


if __name__ == '__main__':
    main()
